background-processor/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+			$node_a.node.handle_funding_signed(&$node_b.node.get_our_node_id(), &get_event_msg!($node_b, MessageSendEvent::SendFundingSigned, $node_a.node.get_our_node_id()));
+			tx
+		}}
+	}
+
+	#[test]
+	fn test_background_processor() {
+		// Test that when a new channel is created, the ChannelManager needs to be re-persisted with
+		// updates. Also test that when new updates are available, the manager signals that it needs
+		// re-persistence and is successfully re-persisted.
+		let nodes = create_nodes(2, "test_background_processor".to_string());
+
+		// Initiate the background processors to watch each node.
+		let _processor_0 = BackgroundProcessor::start(nodes[0].persister.get_data_dir(), FilesystemPersister::persist_manager, nodes[0].node.clone(), nodes[0].logger.clone());
+		let _processor_1 = BackgroundProcessor::start(nodes[1].persister.get_data_dir(), FilesystemPersister::persist_manager, nodes[1].node.clone(), nodes[1].logger.clone());
+
+		// Go through the channel creation process until each node should have something persisted.
+		let tx = open_channel!(nodes[0], nodes[1], 100000);
+
+		macro_rules! check_persisted_data {
+			($node: expr, $filepath: expr) => {
+				let bytes = loop {
+					match std::fs::read($filepath) {
+						Ok(bytes) => break bytes,
+						Err(_) => continue
+					}
+				};
+				let mut expected_bytes = Vec::new();
+				assert!($node.write(&mut expected_bytes).is_ok());
+				assert_eq!(bytes, expected_bytes);
+				// Check that the ChannelManager reset the persistence lock's bool to
+				// false after signalling for updates.
+				let mutcond = Arc::clone(&$node.persistence_notifier.persistence_lock);
+				let &(ref mtx, _) = &*mutcond;
+				let guard = mtx.lock().unwrap();
+				assert_eq!(*guard, false);
+				mem::drop(guard);
+			}
+		}
+
+		// Check that the initial channel manager data is persisted as expected.
+		let filepath_0 = util::get_full_filepath("test_background_processor_persister_0".to_string(), "manager".to_string());
+		let filepath_1 = util::get_full_filepath("test_background_processor_persister_1".to_string(), "manager".to_string());
+		check_persisted_data!(nodes[0].node, filepath_0.clone());
+		check_persisted_data!(nodes[1].node, filepath_1.clone());
+
+		// Save the current bytes then trigger another update.
+		let current_bytes = std::fs::read(filepath_0.clone()).unwrap();
+
+		// Force-close the channel.
+		nodes[0].node.force_close_channel(&OutPoint { txid: tx.txid(), index: 0 }.to_channel_id());
+
+		// Loop until what's written on disk changes.
+		loop {
+			match std::fs::read(filepath_0.clone()) {
+				Ok(bytes) => {
+					if bytes == current_bytes { continue } else { break }
+				},
+				Err(_) => continue
+			}
+		};
+		// Make sure the update was persisted properly.
+		check_persisted_data!(nodes[0].node, filepath_0.clone());
+		// TODO(val): `Drop` is never called on the FilesystemPersisters (causing their directories to linger). why?
```

Comment:
It may be that `Drop` needs to be defined in this crate's test module as well. Perhaps it could be defined outside the other crate's test module only conditionally.